src/lightning/pytorch/loops/fit_loop.py

```diff
@@ -244,12 +244,28 @@ def setup_data(self) -> None:
         if trainer.datamodule is not None:
             allow_zero_length |= trainer.datamodule.allow_zero_length_dataloader_with_multiple_devices
 
+        limits = []
+        for dl in combined_loader.flattened:
+            # determine number of batches
+            length = len(dl) if has_len_all_ranks(dl, trainer.strategy, allow_zero_length) else float("inf")
+            num_batches = _parse_num_batches(stage, length, trainer.limit_train_batches)
+            limits.append(num_batches)
+
+        combined_loader.limits = limits
+
+        training = trainer.training
+        trainer.training = True
+        self._data_fetcher = _select_data_fetcher(trainer)
+        trainer.training = training
+
+        self._data_fetcher.setup(combined_loader)
+        iter(self._data_fetcher)  # creates the iterator inside the fetcher
+        max_batches = sized_len(combined_loader)
+        self.max_batches = max_batches if max_batches is not None else float("inf")
         has_len_all_ranks_ = has_len_all_ranks(combined_loader, trainer.strategy, allow_zero_length)
```

Comment:
I don't think this works anymore. Because with your changes, combined_loader will always return a length (float if undefined).

One possible option is that you take the "or" of all the values across the loop above: https://github.com/Lightning-AI/lightning/pull/18441/files#diff-606e20ddf8a21176c31fe577a4213b0ad569913581b6ae29155b712991cdaa05R248

Reply:
No I don't follow your argument. The combined loader can't return float. While it the return statement inside `__len__` may contain a float value, Python will immediately raise a TypeError. This is then caught inside the `has_len_all_ranks` by the call to `sized_len`.